src/jdk.httpserver/share/classes/sun/net/httpserver/ServerImpl.java

```diff
@@ -382,8 +402,21 @@ void addEvent (Event r) {
     class Dispatcher implements Runnable {
 
         private void handleEvent (Event r) {
+
+            // Stopping marking the state as finished if stop is requested,
+            // termination is in progress and exchange count is 0
+            if (r instanceof StopRequestedEvent) {
+                logger.log(Level.TRACE, "Stop event requested");
+
+                if (terminating && getExchangeCount() == 0) {
```

Comment:
It looks like we may have to check `reqConnection.isEmpty()` here too for the case where the `ServerImpl` is configured with an executor, as it looks like the `exchangeCount` will be incremented in the executor. It would be interesting to try & craft a test with that scenario - possibly by configuring the server with an executor that sleeps some before calling `run()`, thus giving the oportunity of the `StopRequestedEvent` to be handled before the `ExchangeImpl` is created.

Also if we reach here we are terminating - so we could simply assert that.

```suggestion
                boolean terminating = this.terminating;
                assert terminating;
                if (getExchangeCount() == 0 && reqConnection.isEmpty()) {
```

I suspect the other place where we set `finished = true` should also check that `reqConnection.isEmpty()` (and there we do need to check terminating).

Reply:
Done in the next commit